cuda_core/tests/test_program.py

```diff
@@ -150,3 +227,105 @@ def test_program_close():
     program = Program(code, "c++")
     program.close()
     assert program.handle is None
+
+
+
+
+def test_nvvm_deferred_import():
+    """Test that our deferred NVVM import works correctly"""
+    try:
+        from cuda.core.experimental._program import _get_nvvm_module
+        nvvm = _get_nvvm_module()
+        assert nvvm is not None
+    except ImportError as e:
+        pytest.skip(f"NVVM not available: {e}")
+
+
+def test_nvvm_program_creation():
+    """Test basic NVVM program creation"""
+    try:
+        program = Program(nvvm_ir, "nvvm")
+        assert program.backend == "NVVM"
+        assert program.handle is not None
+        program.close()
+    except ImportError as e:
+        pytest.skip(f"NVVM not available: {e}")
+
+
+def test_nvvm_compile_invalid_target():
+    """Test that NVVM programs reject invalid compilation targets"""
+    try:
+        program = Program(nvvm_ir, "nvvm")
+        with pytest.raises(ValueError, match="NVVM backend only supports target_type=\"ptx\""):
+            program.compile("cubin")
+        program.close()
+    except ImportError as e:
+        pytest.skip(f"NVVM not available: {e}")
+
+
+def test_nvvm_compile_to_ptx(init_cuda):
+    """Test NVVM IR compilation to PTX"""
+    try:
+        options = ProgramOptions(name="nvvm_test", arch="sm_90", device_code_optimize=False)
+        program = Program(nvvm_ir, "nvvm", options=options)
+        try:
+            ptx_object_code = program.compile("ptx")
+            assert isinstance(ptx_object_code, ObjectCode)
+            assert ptx_object_code.name == "nvvm_test"
+            assert ptx_object_code._code_type == "ptx"
+            
+            ptx_code = ptx_object_code.code
+            if isinstance(ptx_code, bytes):
+                ptx_text = ptx_code.decode()
+            else:
+                ptx_text = str(ptx_code)
+            assert ".visible .entry simple(" in ptx_text
+            
+            ptx_kernel = ptx_object_code.get_kernel("simple")
+            assert isinstance(ptx_kernel, Kernel)
+            
+        except Exception as e:
+            if any(error in str(e) for error in ["ERROR_IR_VERSION_MISMATCH", "ERROR_INVALID_OPTION", "ERROR_COMPILATION"]):
+                pytest.skip(f"NVVM IR not compatible with this CUDA version: {e}")
+            else:
+                raise
+        finally:
+            program.close()
+            
+    except ImportError as e:
+        pytest.skip(f"NVVM not available: {e}")
+
+
+@pytest.mark.parametrize("options", [
+    ProgramOptions(name="test1", arch="sm_90", device_code_optimize=False),
+    ProgramOptions(name="test2", arch="sm_90", device_code_optimize=False),
```

Comment:
Q: 
1. What's the difference between these two functions?
2. Isn't this test identical to the previous `test_nvvm_compile_to_ptx` test? Do we need both? Seems like a lot of code dup to me.

Reply:
yeah agree , was planning to remove this . This I initially copied to check with different Sm versions.